server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java

```diff
@@ -1162,6 +1166,83 @@ public void testShardStats() throws IOException {
+        Settings settings = Settings.builder().put(IndexMetaData.SETTING_VERSION_CREATED, Version.CURRENT)
+                .put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, 0)
+                .put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, 1)
+                .build();
+        IndexMetaData metaData = IndexMetaData.builder(shardRouting.getIndexName())
+                .settings(settings)
+                .primaryTerm(0, 1)
+                .build();
+
+        // Override one Directory methods to make it fail at our will
+        // We use AtomicReference here to inject failure in the middle of the test not immediately
+        // We use Supplier<IOException> instead of IOException to produce meaningful stacktrace
+        // (remember stack trace is filled when exception is instantiated)
+        AtomicReference<Supplier<IOException>> exceptionToThrow = new AtomicReference<>();
+        AtomicBoolean throwWhenMarkingStoreCorrupted = new AtomicBoolean(false);
+        Directory directory = new FilterDirectory(newFSDirectory(shardPath.resolveIndex())) {
+            //fileLength method is called during storeStats try block
+            //it's not called when store is marked as corrupted
+            @Override
+            public long fileLength(String name) throws IOException {
+                Supplier<IOException> ex = exceptionToThrow.get();
+                if (ex == null) {
+                    return super.fileLength(name);
+                } else {
+                    throw ex.get();
+                }
+            }
+
+            //listAll method is called when marking store as corrupted
+            @Override
+            public String[] listAll() throws IOException {
+                Supplier<IOException> ex = exceptionToThrow.get();
+                if (throwWhenMarkingStoreCorrupted.get() && ex != null) {
+                    throw ex.get();
+                } else {
+                    return super.listAll();
+                }
+            }
+        };
+
+        Store store = createStore(shardId, new IndexSettings(metaData, Settings.EMPTY), directory);
+        IndexShard shard = newShard(shardRouting, shardPath, metaData, store,
+                null, new InternalEngineFactory(), () -> {}, EMPTY_EVENT_LISTENER);
+
+        recoverShardFromStore(shard);
+        boolean corruptIndexException = true;
+
+        if (randomBoolean()) {
+            exceptionToThrow.set(() -> new CorruptIndexException("Test CorruptIndexException", "Test resource"));
+            throwWhenMarkingStoreCorrupted.set(randomBoolean());
+        } else {
+            corruptIndexException = false;
+            exceptionToThrow.set(() -> new IOException("Test IOException"));
+        }
+        ElasticsearchException e = expectThrows(ElasticsearchException.class, shard::storeStats);
+        assertThat(shard.state(), equalTo(IndexShardState.CLOSED));
```

Comment:
@bleskes You mean, that I can just check if shard failure event listener has triggered instead of checking shard state? If I remove shard closing from implementation, then yes, actually shard state would be opened. It was in original proposed patch by you, that's why I've included logic to close the shard.

Reply:
> @bleskes You mean, that I can just check if shard failure event listener has triggered instead of checking shard state?

yes

>  It was in original proposed patch by you, that's why I've included logic to close the shard.

I told you I wasn't sure of it's quality :)